lib/common/xpath.c

```diff
@@ -266,38 +266,68 @@ get_xpath_object(const char *xpath, xmlNode * xml_obj, int error_level)
     return result;
 }
 
-int
-pcmk__element_xpath(const char *prefix, const xmlNode *xml, char *buffer,
-                    int offset, size_t buffer_size)
+/*!
+ * \internal
+ * \brief Get an XPath string that matches an XML element as closely as possible
+ *
+ * \param[in] xml  The XML element for which to build an XPath string
+ *
+ * \return A \p GString that matches \p xml, or \p NULL if \p xml is \p NULL.
+ *
+ * \note The caller is responsible for freeing the string.
+ */
+GString *
+pcmk__element_xpath(const xmlNode *xml)
 {
-    const char *id = ID(xml);
+    const char *id = NULL;
+    GString *xpath = NULL;
+
+    if (xml == NULL) {
+        return NULL;
+    }
+
+    xpath = pcmk__element_xpath(xml->parent);
+
+    if (TYPE(xml) == NULL) {
```

Comment:
I didn't expect to have to keep this logic, but apparently the top-level XML element is nameless. Found that out the hard way.

Reply:
Comments and text nodes have no name, but we probably don't ever generate xpaths for those (or have a use case for that), so maybe we should return NULL right away if xml isn't an element.

I believe all XML elements have a name so I'm not sure why the top-level element would hit this. Maybe it's actually the XML_DOCUMENT_NODE? But I would expect xml->parent and thus xpath is NULL in those cases.

Reply:
It must have been the document node that was nameless. When I added debugging statements, I got output like `"/(null/cib/configuration/..."`.

Reply:
Assuming xml->parent is NULL for the document node, we should be able to just return early for anything that's not an element, then we don't need the type check

Reply:
That seems to work. If there is somehow a node that has a parent and doesn't have a name, we're going to end up with a weird/invalid XPath no matter what. So I don't think we need to worry about that right now.